modules/imgcodecs/src/loadsave.cpp

```diff
@@ -1032,6 +923,187 @@ bool haveImageWriter( const String& filename )
     return !encoder.empty();
 }
 
+class ImageCollection::Impl {
+public:
+    Impl() = default;
+    Impl(const std::string&  filename, int flags);
+    void init(String const& img, int flags);
+    size_t size() const;
+    Mat read();
+    int width() const;
+    int height() const;
+    bool readHeader();
+    Mat readData();
+    bool advance();
+    int currentIndex() const;
+
+private:
+    String m_filename;
+    int m_flags{};
+    std::size_t m_size{};
+    int m_width{};
+    int m_height{};
+    int m_current{};
+    ImageDecoder m_decoder;
+};
+
+ImageCollection::Impl::Impl(std::string const& filename, int flags) {
+    this->init(filename, flags);
+}
+
+void ImageCollection::Impl::init(String const& filename, int flags) {
+    m_filename = filename;
+    m_flags = flags;
+
+#ifdef HAVE_GDAL
+    if (m_flags != IMREAD_UNCHANGED && (m_flags & IMREAD_LOAD_GDAL) == IMREAD_LOAD_GDAL) {
+        m_decoder = GdalDecoder().newDecoder();
+    }
+    else {
+#endif
+    m_decoder = findDecoder(filename);
+#ifdef HAVE_GDAL
+    }
+#endif
+
+
+    if (!m_decoder) {
+        throw cv::Exception(1, "No Encoder found for the file named:" + m_filename,
+                            String("ImageCollection::Constructor"),
+                            __FILE__,
+                            __LINE__);
+    }
+
+    m_decoder->setSource(filename);
+
+    if (!m_decoder->readHeader())
+        throw cv::Exception(1, "Can not read image header in ImageCollection" +
+                                String(" file named " + m_filename),
+                                String("ImageCollection::Constructor"),
+                                __FILE__,
+                                __LINE__);
+
+    size_t count = 1;
+    while(m_decoder->nextPage()) count++;
```

Comment:
This `nextPage` call is required to find the ImageCollection size. About extracting all position dependent information, I need to pass `m_decoder` and `this` to the iterator in order to do that. So, i am not sure if this is a good idea because operator[] and at() also decodes pages. `this` is required for `operator*` decoding. So it can put pages inside the `m_vector`. I think we need to discuss this.